spark/src/test/scala/org/apache/spark/sql/delta/IdentityColumnAdmissionSuite.scala

```diff
@@ -0,0 +1,267 @@
+/*
+ * Copyright (2021) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.delta
+
+import java.io.{FileNotFoundException, PrintWriter}
+
+import org.apache.spark.sql.delta.GeneratedAsIdentityType.{GeneratedAlways, GeneratedAsIdentityType}
+import org.apache.spark.sql.delta.actions.RemoveFile
+import org.apache.spark.sql.delta.sources.{DeltaSourceUtils, DeltaSQLConf}
+import org.apache.hadoop.fs.Path
+
+import org.apache.spark.{SparkConf, SparkException}
+import org.apache.spark.sql.{AnalysisException, Row}
+import org.apache.spark.sql.catalyst.TableIdentifier
+import org.apache.spark.sql.catalyst.expressions.Cast
+import org.apache.spark.sql.catalyst.parser.ParseException
+import org.apache.spark.sql.execution.streaming.MemoryStream
+import org.apache.spark.sql.streaming.{StreamingQueryException, Trigger}
+import org.apache.spark.sql.types.{DoubleType, IntegerType, LongType}
+import org.apache.spark.util.Utils
+
+// Test command that should be allowed and disallowed on IDENTITY columns.
+trait IdentityColumnAdmissionSuiteBase
+  extends IdentityColumnTestUtils {
+
+  import testImplicits._
+
+  protected val tblName = "identity_admission_test"
+
+  protected override def sparkConf: SparkConf = {
+    super.sparkConf
+      .set(DeltaSQLConf.DELTA_IDENTITY_COLUMN_ENABLED.key, "true")
+  }
+
+  test("alter table change column type") {
+    for {
+      generatedAsIdentityType <- GeneratedAsIdentityType.values
+      keyword <- Seq("ALTER", "CHANGE")
+      targetType <- Seq(IntegerType, DoubleType)
+    } {
+      withIdentityColumnTable(generatedAsIdentityType, tblName) {
+        // Type changes that aren't upcast are rejected early during analysis by Spark,
+        // while upcasts are rejected in Delta when altering data type of an identity column.
+        if (Cast.canUpCast(LongType, targetType)) {
+          val ex = intercept[DeltaAnalysisException] {
+              sql(s"ALTER TABLE $tblName $keyword COLUMN id TYPE ${targetType.sql}")
+            }
+          assert(ex.getErrorClass === "DELTA_IDENTITY_COLUMNS_ALTER_COLUMN_NOT_SUPPORTED")
+        } else {
```

Comment:
can we do a case match on `targetType` instead of this?

```
targetType match {
   case IntegerType => // foo
   case DoubleType => // bar
   case _ => fail("unexpected targetType")
}

```